### PR TITLE
Add watch expressions by context menu in editor (#1596)

### PR DESCRIPTION
**Pull request by jbhoosreddy**
Associated Issue: #1596 

### Summary of Changes

* Add context menu when string is selected in editor to add watch expression

### CODE REVIEW

**Comment by jbhoosreddy** [comment]
@jasonLaster @clarkbw let me know if there are any code changes / UI changes that could be made to improve the original PR. Thanks!

File: src/actions/pause.js [hunk 1]
```diff
@@ -220,8 +220,14 @@ function loadObjectProperties(grip: Grip) {
  */
 function addExpression(expression: Expression) {
   return ({ dispatch, getState }: ThunkArgs) => {
+    const expressions = getExpressions(getState()).toSeq();
+    const filteredExpressions = expressions
+      .filter(e => e.input === expression.input);
+    if (expressions.size && filteredExpressions.count()) {
+      return;
+    }
```
Comment (jasonLaster) [inline src/actions/pause.js]: Hmm, perhaps we use the `includes` function

```js
function expressionExists() {
   return !!expressions.find(e => ..)
}
```

```js
if (expressionExists(expression, expressions)) {
  return
}
```

File: assets/locales/debugger.properties [hunk 2]
```diff
@@ -247,6 +247,10 @@ editor.conditionalPanel.placeholder=This breakpoint will pause when the expressi
 # for navigating to a source mapped location
 editor.jumpToMappedLocation=Jump to %s location
 
+# LOCALIZATION NOTE(editor.watchExpression): Context menu item
+# for adding watch expression from Editor
+editor.watchExpression=Watch this expression
```
Comment (jasonLaster) [inline assets/locales/debugger.properties]: @clarkbw 

* "Add Watch Expression"
* "Add selected text to watches"

File: src/components/Editor.js [hunk 3]
```diff
@@ -127,9 +128,23 @@ const Editor = React.createClass({
...
     };
 
-    showMenu(event, [
+    const menuOptions = [
       jumpLabel
-    ]);
+    ];
+
+    if (this.editor.codeMirror.somethingSelected()) {
+      const watchExpressionLabel = {
```
Comment (jasonLaster) [inline src/components/Editor.js]: NIT: lets create this label regardless and only do the push if the condition is true

File: src/components/Editor.js [hunk 4]
```diff
@@ -127,9 +128,23 @@ const Editor = React.createClass({
...
+        accessKey: "E",
+        disabled: false,
+        label: L10N.getStr("editor.watchExpression"),
+        click: () => this.props.addExpression({
```
Comment (jasonLaster) [inline src/components/Editor.js]: lets move this to an instance function as we probably want to also check to see if the cursor is in the selection

Comment (jbhoosreddy) [inline src/components/Editor.js]: @jasonLaster any choice on the accessKey?

Comment (jbhoosreddy) [inline src/components/Editor.js]: > lets move this to an instance function as we probably want to also check to see if the cursor is in the selection

couldn't find a straightforward way of doing it. basically, the cursor stays within the selection even though the pointer is right-clicking somewhere else. will follow-up on this in a subsequent PR.

File: src/actions/pause.js [hunk 5]
```diff
@@ -223,13 +223,18 @@ function loadObjectProperties(grip: Grip) {
...
     if (!expression.input) {
       return;
     }
 
+    const expressions = getExpressions(getState()).toSeq();
+    const matchedExpression = expressions
+      .find(e => e.input === expression.input);
+    if (matchedExpression !== undefined) {
+      return;
+    }
```
Comment (jasonLaster) [inline src/actions/pause.js]: I think we can clean this up: 

```js
function expressionExists() {
   return !!expressions.find(e => ..)
}
```

```js
if (expressionExists(expression, expressions)) {
  return
}
```

File: src/actions/pause.js [hunk 6]
```diff
@@ -223,13 +223,18 @@ function loadObjectProperties(grip: Grip) {
...
+    const id = expression.id !== undefined ? parseInt(expression.id, 10) :
+      expressions.size++;
```
Comment (jasonLaster) [inline src/actions/pause.js]: why is this more complicated than before?

File: src/components/Editor.js [hunk 7]
```diff
@@ -127,9 +128,21 @@ const Editor = React.createClass({
...
+    const watchExpressionLabel = {
+      accesskey: "E",
+      disabled: !this.editor.codeMirror.somethingSelected(),
+      label: L10N.getStr("expressions.placeholder"),
+      click: () => this.props.addExpression({
+        input: this.editor.codeMirror.getSelection()
+      })
+    };
+
+    const menuOptions = [
```
Comment (jasonLaster) [inline src/components/Editor.js]: Hmm, we should probably still have the feature flag

```js
if (isEnabled)
  menuOptions.push(watchExpressionLabel)
```

**Comment by jasonLaster** [comment]
@jbhoosreddy updated, let me know what you think

**Comment by jasonLaster** [comment]
@wldcordeiro merged it as you're traveling. Let me know if you have any comments and we can easily do a follow up.

**Comment by jbhoosreddy** [comment]
No it looks great! I like it!

**Pull request merged**